Make bootstrap admins temporary conditionally

slaskawi opened the pull request:
This Pull Request introduces a flag to make the Bootstrap Admin temporary or permanent. 

Closes #40305

slaskawi committed:
Make bootstrap admins temporary conditionally

Signed-off-by: Sebastian Łaskawiec <sebastian.laskawiec@defenseunicorns.com>

shawkins reviewed (changes requested):
Thanks @slaskawi, just a couple of things that could be addressed.

slaskawi commented on services/src/main/java/org/keycloak/services/managers/ApplianceBootstrap.java:
```diff
@@ -121,7 +121,7 @@ public boolean createMasterRealm() {
      * @return false if the user could not be created
      */
-    public boolean createTemporaryMasterRealmAdminUser(String username, String password, /*Integer expriationMinutes,*/ boolean initialUser) {
+    public boolean createTemporaryMasterRealmAdminUser(String username, String password, boolean isTemporary, /*Integer expriationMinutes,*/ boolean initialUser) {
```
Fixed!

slaskawi commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/command/BootstrapAdminUser.java:
```diff
@@ -54,6 +54,12 @@ static class UsernameOptions {
+    @Option(paramLabel = "IS_TEMPORARY", names = { "--is-temporary" }, description = "Indicates whether this admin account is temporary, defaults to true", defaultValue = "true")
+    boolean isTemporary;
+
+    @Option(paramLabel = "is_temporary", names = { "--is-temporary:env" }, description = "Environment variable name for whether the admin account is temporary")
```
Removed!

slaskawi commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testExportHelp.approved.txt:
```diff
@@ -261,6 +261,10 @@ Bootstrap Admin:
 --bootstrap-admin-password <password>
                      Temporary bootstrap admin password. Used only when the master realm is
```
That's a good catch! I updated the descriptions and removed the term "temporary". Should read better now.

slaskawi commented on services/src/main/java/org/keycloak/services/resources/WelcomeResource.java:
```diff
@@ -302,4 +305,14 @@ private void csrfCheck(final MultivaluedMap<String, String> formData) {
         }
     }
 
+    private boolean createsTemporaryAdminUser() {
```
> Also behavior doesn't match the behavior of the other properties. If I set the admin username env variable, that does not become the default in the form.

At the moment, this works on two levels - it can be set via an Environment Property (`KC_BOOTSTRAP_ADMIN_IS_TEMPORARY`) or in the `theme.properties` file. The latter gives a chance of override this in a custom theme. At the moment, the theme override takes the priority over the environment variable to give custom theme authors more flexibility.

I'm open for a discussion if you'd like me to change the mechanism how this works.

vmuzikar replied:
> in particular the usage of the environment variable since it's not consistent behavior

Exactly.

Additionally, I do not believe a theme prop is something that should affect this, it's not its responsibility, IMHO.

I'm ok with having the ability to set the temporary flag via the Welcome Screen, though. However, we should use a systematic approach: fully leverage the `bootstrap-admin-temporary` option, i.e. it would accept the env var, the CLI option, etc.

slaskawi replied:
Adjusted

slaskawi commented:
@shawkins The comments have been addressed and the change is ready for the next round.

The major changes are in Welcome Page theme. It now doesn't print information about temporary user if the `kcBootstrapAdminIsTemporary` setting is set to `false`.

I'm still considering writing a unit test for it but this PR is also fine without it (this is a bit niche feature).

shawkins commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/jaxrs/QuarkusKeycloakApplication.java:
```diff
@@ -77,16 +78,17 @@ protected void loadConfig() {
     protected void createTemporaryAdmin(KeycloakSession session) {
         var adminUsername = getOption(BootstrapAdminOptions.USERNAME.getKey(), KEYCLOAK_ADMIN_ENV_VAR);
         var adminPassword = getOption(BootstrapAdminOptions.PASSWORD.getKey(), KEYCLOAK_ADMIN_PASSWORD_ENV_VAR);
+        var isTemporary = Boolean.parseBoolean(Configuration.getOptionalKcValue(BootstrapAdminOptions.EXPIRATION.getKey()).orElse("true"));
```
This should not reference IS_TEMPORARY correct? 

The EXPIRATION option was never implemented - it still might at some point, which will cause some additional validation with the IS_TEMPORARY flag.

slaskawi replied:
You're eagle-eyed @shawkins ! Thanks for spotting this!

shawkins commented on services/src/main/java/org/keycloak/services/resources/WelcomeResource.java:
```diff
@@ -302,4 +310,14 @@ private void csrfCheck(final MultivaluedMap<String, String> formData) {
 
+    protected boolean createsTemporaryAdminUser() {
+        try {
+            String themeProperty = "kcBootstrapAdminIsTemporary";
```
The method name should be createAdminUser, right?

slaskawi replied:
I renamed it to `isAdminUserTemporary` to make it more descriptive. In general, this returns `true` if the Welcome Page should create a temporary user.

slaskawi commented on docs/guides/server/bootstrap-admin-recovery.adoc:
```diff
@@ -105,4 +111,6 @@ For the `bootstrap-admin service` command, the client ID is optional and default
 
 <@kc.bootstrapadmin parameters="service --client-id:env <YourClientIdEnv> --client-secret:env <YourSecretEnv>"/>
 
+Both commands also support using the `--is-temporary:env` parameter and enables using a different environment variable to control whether the created account is temporary or permanent.
```
Done.

vmuzikar commented on docs/guides/server/bootstrap-admin-recovery.adoc:
```diff
@@ -95,6 +95,12 @@ The `--no-prompt` parameter can be useful if the username or client ID should be
+Both `bootstrap-admin user` and `bootstrap-admin service` commands support the `--is-temporary` parameter. This parameter allows you to treat the created admin account as a permanent one, meaning it will not be marked as temporary and will not be subject to the temporary account restrictions.
+
+A temporary admin account can also be created with the Welcome Screen that is displayed at the first start of {project_name}. In this case, the temporary account behavior is controlled either by the `KC_BOOTSTRAP_ADMIN_IS_TEMPORARY` environment variable or by the `kcBootstrapAdminIsTemporary` setting in `theme.properties`. If the variable or setting is not defined, the account will be created as  ...
+
```
We should probably add some clarification on what we mean by "temporary". Perhaps something along these lines:
```suggestion

At the moment, a temporary account is reflected only in a form of visual warnings when using it, it does not actually expire the account. This might change with a future release.
```

vmuzikar reviewed (changes requested):
@slaskawi Thank you for the contribution!

slaskawi commented:
@vmuzikar @shawkins Thanks for the comments. I've pushed another update, here are the highlights:

- Removed the environment variable handling in `WelcomeResource` as suggested
- Unified approach with bootstrap parameters
- Updated the documentation

slaskawi commented on quarkus/config-api/src/main/java/org/keycloak/config/BootstrapAdminOptions.java:
```diff
@@ -5,17 +5,23 @@ public class BootstrapAdminOptions {
+    // Since the Configuration object is not available in the WelcomeResource, we'll just pass it through a static
+    // handler. It's definitely a suboptimal approach but the alternatives are to either put it in the Profile
+    // (which doesn't feel right) or RealmModel (which seems like too much).
+    private static boolean isTemporaryRuntimeAdmin = true;
```
@shawkins @vmuzikar Unfortunately I didn't find any better way to read the `--is-temporary` bootstrap parameter value from the `WelcomeResource`. This approach is not pretty but seems to work correctly. I'm open for suggestions on how to improve it (or we can proceed as is if it's OK).

shawkins replied:
Is it truly necessary for the is-temporary option to influnce the WelcomeResource?

The alternative to this static variable to map the is-temporary option to an spi key and use Config.scope

We mused about providing the root configuration to Config via something like https://github.com/keycloak/keycloak/pull/39123 but it didn't move forward.

vmuzikar replied:
> The alternative to this static variable to map the is-temporary option to an spi key and use Config.scope

Yes, that would be the correct approach.

shawkins commented on services/src/main/java/org/keycloak/services/resources/WelcomeResource.java:
```diff
@@ -302,4 +310,16 @@ private void csrfCheck(final MultivaluedMap<String, String> formData) {
         }
     }
 
+    protected String getThemeProperty(String propertyName, String defaultValue) {
```
This is a minor behavioral change to the existing logic - catching rather than throwing an IOException. Is this method for subclasses? And / or does it be long as a utility method on Theme?

vmuzikar commented on docs/guides/server/bootstrap-admin-recovery.adoc:
```diff
@@ -95,6 +95,13 @@ The `--no-prompt` parameter can be useful if the username or client ID should be
+
+Both `bootstrap-admin user` and `bootstrap-admin service` commands support the `--is-temporary` parameter. This parameter allows you to treat the created admin account as a permanent one, meaning it will not be marked as temporary and will not be subject to the temporary account restrictions.
+
+A temporary admin account can also be created with the Welcome Screen that is displayed at the first start of {project_name}. In this case, the temporary account behavior is controlled by the `--is-temporary` parameter. If the variable or setting is not defined, the account will be created as a temporary one.
```
This is not quite correct, is it? For the start command, the option is named `bootstrap-admin-temporary`. We should mention it more in general that the temporary flag is controlled not only by `--is-temporary`.

slaskawi replied:
You're absolutely right! Corrected.

vmuzikar commented on quarkus/config-api/src/main/java/org/keycloak/config/BootstrapAdminOptions.java:
```diff
@@ -27,13 +33,28 @@ public class BootstrapAdminOptions {
+
+    public static final Option<Boolean> IS_TEMPORARY = new OptionBuilder<>("bootstrap-admin-temporary", Boolean.class)
+            .category(OptionCategory.BOOTSTRAP_ADMIN)
+            .description("Indicates whether the admin user or service account is temporary." + USED_ONLY_WHEN + NON_CLI)
```
Why would we suggest the CLI, it's not a secret.
```suggestion
            .description("Indicates whether the admin user or service account is temporary." + USED_ONLY_WHEN)
```

vmuzikar reviewed (changes requested):
@slaskawi Thank you for the updates.

slaskawi commented:
> Is it truly necessary for the is-temporary option to influnce the WelcomeResource?
> The alternative to this static variable to map the is-temporary option to an spi key and use Config.scope
> We mused about providing the root configuration to Config via something like https://github.com/keycloak/keycloak/pull/39123 but it didn't move forward.

For the UDS Platform, having this functionality in WelcomeResource is vital. Putting this into Config seems a bit artificial as this is not really an SPI and there's only 1 implementation that probably won't be overridden. Maybe I'll just add a link to https://github.com/keycloak/keycloak/pull/39123 in the comment so that we can retrofit it later once we have better options.

shawkins commented on docs/guides/server/bootstrap-admin-recovery.adoc:
```diff
@@ -95,6 +95,13 @@ The `--no-prompt` parameter can be useful if the username or client ID should be
+
+Both `bootstrap-admin user` and `bootstrap-admin service` commands support the `--is-temporary` parameter. This parameter allows you to treat the created admin account as a permanent one, meaning it will not be marked as temporary and will not be subject to the temporary account restrictions.
+
+A temporary admin account can also be created with the Welcome Screen that is displayed at the first start of {project_name}. In this case, the temporary account behavior is controlled by the `--bootstrap-admin-temporary` parameter. If the variable or setting is not defined, the account will be created as a temporary one.
```
@vmuzikar that wouldn't be consistent with the bootstrap admin commands existing handling. The commands do not consult the bootstrap-admin=* options. Someone could be setting bootstrap-admin-* options in keycloak.conf for example, and we don't want that influencing the bootstrap admin command unintentionally.

So --bootstrap-admin-temporary should just affect the initial admin creation / Welcome Screen.

vmuzikar replied:
> So --bootstrap-admin-temporary should just affect the initial admin creation / Welcome Screen.

Yes, that's what I meant. Sorry, I was probably not clear before.

I meant that `--bootstrap-admin-temporary` should also affect when you create an admin user/service account at startup, not just the Welcome Screen. And that this behaviour should be documented.

stianst commented:
I disagree with this approach; we don't have a mechanism to auto-delete the admin account at this point, so this option is complete unnecessary. If/when we do some form of auto removal of the admin account it should include a consideration of the scenarios where someone doesn't want to delete the account, but rather configure it properly.

I propose a completely different approach to the temporary admin account (never been a big fan of this in the first place). See https://github.com/keycloak/keycloak/discussions/41382

stianst reviewed (changes requested):
I'm not in favour of adding this option, see my comment
